python/ql/lib/semmle/python/frameworks/data/ModelsAsData.qll

```diff
@@ -0,0 +1,46 @@
+/**
+ * Provides classes for contributing a model, or using the interpreted results
+ * of a model represented as data.
+ *
+ * - Use the `ModelInput` module to contribute new models.
+ * - Use the `ModelOutput` module to access the model results in terms of API nodes.
+ *
+ * The package name refers to a Pip package name.
```

Comment:
What the package field of the CSV should be used for had a bit of discussion in the EDR, but we did not reach a conclusion. Me and @asgerf decided to not do the discussion in the EDR, so the details are now gone, so let me try to capture it here.

Currently a row for typemodel looks like

```csv
asyncpg;ConnectionPool;asyncpg;;Member[create_pool].ReturnValue.Awaited
```

which corresponds to this Python code

```py
import asyncpg

pool = await asyncpg.create_pool(...)
```

All good and well so far.

But, some packages provide module with different names. For example, the popular [PyYAML](https://pypi.org/project/PyYAML/) PyPI package is used like

```py
import yaml

yaml.unsafe_load(...)
```

and it would be natural to model the sink as

```csv
yaml;;Member[unsafe_load].Argument[0];unsafe-deserialization
```

but now the 'package' part of the CSV no longer corresponds to a PyPI package, but how you import it.

### Using PyPI package name for package field

As I recall it (but do correct me if I'm wrong), @asgerf pointed out that it might be desirable to actually have the PyPI package as the package field, so you could easily know from the MaD rows what PyPI packages we have data for. That means you would end up with something like the following (where I invented a top-level `Import[<name>]` token)

```csv
PyYAML;;Import[yaml].Member[unsafe_load].Argument[0];unsafe-deserialization
```

to me that does not seem like a good solution, since the notation will be different from other languages, and coming from also writing MaD in those, probably frustratingly non-intuitive.

Reply:
I originally wanted `package` to be the package manager's name for the package, as pointed out here I think it will work better to use the name as it appears in the source code.

@RasmusWL does `import foo.bar` currently correspond to `API::moduleImport("foo").getMember("bar")`?

Reply:
> @RasmusWL does import foo.bar currently correspond to `API::moduleImport("foo").getMember("bar")`?

yes :+1:

Reply:
> to me that does not seem like a good solution, since the notation will be different from other languages, and coming from also writing MaD in those, probably frustratingly non-intuitive. 

Agreed, I think we should keep it as the name appears in the source code.  
And I think It should be easy to take an example client and make a model from that, and that is easy when you use the package name from the source.  

> import foo.bar 

I added a test that confirms that you can use `foo.bar` as the package name.

Reply:
> I added a test that confirms that you can use `foo.bar` as the package name.

I was surprised by this :flushed: I thought the package field of the CSV would be reserved for a top-level package, such that the following CSV would be disallowed

```
"foo1.bar;;Member[baz1].Argument[any];test-source"
```

and should always be written as 

```
"foo1;;Member[bar].Member[baz1].Argument[any];test-source"
```

I guess this is a bit of a nitpick, but I think it provides two nice things:

1. there is **one** way to write such a line. Having a canonical representation is desirable from my point of view (just as a general observation). The only application I can see straight off the bat is being able to spot duplicates more easily.
2. if a MaD writing person moves to QL writing, they will know to write `API::moduleImport("foo").getMember("bar")` and won't try to write `API::moduleImport("foo.bar")` which won't work.

Reply:
We never actually updated this wording after agreeing on the new semantics. How about this?

```suggestion
 * The package name refers to the top-level module the import comes from, and not a PyPI package.
 * So for `from foo.bar import baz`, the package will be `foo`.
```

(notice that pip is _one_ of the package managers in the Python ecosystem that uses packages from https://pypi.org/)